src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,356 @@
+        private readonly IBroadcasterManager broadcasterManager;
+        private readonly IFederatedPegSettings settings;
+        private readonly ILogger logger;
+        private readonly Network network;
+        private readonly IAsyncProvider asyncProvider;
+
+        /// <summary>
+        /// Queue to handle incoming blocks and update state if needed.
+        /// </summary>
+        private readonly IAsyncDelegateDequeuer<ChainedHeaderBlock> blockQueue;
+
+        /// <summary>
+        /// The task that is building all the consolidation transactions, if one is currently running.
+        /// </summary>
+        private Task consolidationTask;
+
+        /// <summary>
+        /// Used to protect ConsolidationTransactions from write operations.
+        /// </summary>
+        private readonly object lockObj = new object();
+
+        /// <inheritdoc />
+        public List<ConsolidationTransaction> ConsolidationTransactions { get; private set; }
+
+        // TODO: Could put a dictionary by OutPoint.
+
+        public InputConsolidator(IFederationWalletTransactionHandler transactionHandler,
+            IFederationWalletManager walletManager,
+            IBroadcasterManager broadcasterManager,
+            IFederatedPegSettings settings,
+            ILoggerFactory loggerFactory,
+            ISignals signals,
+            IAsyncProvider asyncProvider,
+            Network network)
+        {
+            this.transactionHandler = transactionHandler;
+            this.walletManager = walletManager;
+            this.broadcasterManager = broadcasterManager;
+            this.network = network;
+            this.settings = settings;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.asyncProvider = asyncProvider;
+            this.blockQueue = asyncProvider.CreateAndRunAsyncDelegateDequeuer<ChainedHeaderBlock>($"{nameof(InputConsolidator)}-{nameof(this.blockQueue)}", this.ProcessBlockInternal);
+            signals.Subscribe<WalletNeedsConsolidation>(this.StartConsolidation);
+        }
+
+        /// <inheritdoc />
+        public void StartConsolidation(WalletNeedsConsolidation trigger)
+        {
+            // If there isn't already a task running, start it.
+
+            if (this.consolidationTask == null || this.consolidationTask.IsCompleted)
+            {
+                this.consolidationTask = Task.Run(() =>
+                {
+                    lock (this.lockObj)
+                    {
+                        // If we're already in progress we don't need to do anything.
+                        if (this.ConsolidationTransactions != null)
+                            return;
+
+                        this.logger.LogInformation("Building consolidation transactions for federation wallet inputs.");
+
+                        this.ConsolidationTransactions = this.CreateRequiredConsolidationTransactions(trigger.Amount);
```

Comment:
`CreateRequiredConsolidationTransactions` is called under a lock and inside acquires a lock again.
Also I think it shouldn't be public since in production code it's only called from the consolidator. 

Please remove it from the public interface, remove squiring the lock inside and add `Locked` to the method name + a remarks comment like 'should be protected by see=lockobj'.

Reply:
Hmm it's really useful for us to be able to test though - acquiring a lock inside of a lock isn't a problem right?

Reply:
`acquiring a lock inside of a lock isn't a problem right?` No, it's not. 
The problem I see is that private method is exposed on the public interface just in favor of tests. This method can be tested via reflection\via other methods indirectly

Reply:
Yeah fair, I agree, I'd prefer not to change it now though, it will just be a pain to test.

Reply:
Sure, I'm ok with it if you will change it in the next PR

Reply:
Cool I can add exit logs and change this both in a separate PR soon